### ISSUE TITLE
[Feature Request] Low latency output on Windows via ASIO/WASAPI exclusive

### ISSUE THREAD

**Issue by ThreeDeeJay** [kcat/openal-soft#682]
High audio latency (100ms+) is something that has plagued apps/games on Windows for a long time, yet people rarely notice, let alone measure it, like [Matt Gore, HeSuVi developer](https://sourceforge.net/p/hesuvi/wiki/Audio%20Latency%20Measurements/) and [Battle(non)sense](https://youtu.be/Kggh_Ge8YHs?t=352).
![image](https://i.imgur.com/ZKy640S.jpeg)
![image](https://i.imgur.com/aefWeum.jpeg)

So I've been wondering if ASIO could be implemented into OpenAL Soft directly, since [Crystal Mixer already did something like that](https://sourceforge.net/projects/crystal-mixer/files/Crystal%20Mixer/), but AFAIK it's only capable of virtualizing the multichannel audio mix.

Alternatively, someone even modified OpenAL Soft to use WASAPI in exclusive mode, which I've tested and confirmed it does make a difference (tho I'm yet to measure it), so I forked it [here](https://github.com/kcat/openal-soft/compare/master...ThreeDeeJay:openal-soft:WASAPI-Exclusive)
Perhaps a flag to switch to exclusive mode in the main branch would be more feasible since it just seems to require a couple line edits (tho it would probably need some improvement so it's not restricted to `sample-type=int16` and `period_size` is automatically set to the lowest supported by the sound card, as well as minimal/no mixahead or any other bottlenecks to ensure lowest possible latency).

Either option should hopefully allow ultra low latency on [thousands of games that are at least potentially supported by OpenAL Soft](https://airtable.com/shryB7ft0Cle17cLA) by using the sound card's native ASIO or ASIO4ALL. I think audio would also be bit-perfect audio (at least on WASAPI exclusive) by bypassing the Windows mixer. So perhaps eventually including both would give people options based on their needs.

**Comment by kcat** [comment]
I'm curious how much of the latency is a result of using shared mode or non-ASIO output. "Button to audio" latency with random games doesn't say much, since it's also including input latency (the time from physically pressing a button to the OS detecting the input, then to the process detecting the input), and logic/frame latency (the time from the process getting input to processing a new logic frame, and from a logic frame to updating audio state, which can be at different rates), and only then getting to the audio latency. 

OpenAL Soft itself will add about 50ms on average, given the default 20ms period size and 60ms buffer. Certain post processors may add a couple more milliseconds (output limiter, UHJ encoder, etc, which will be reported as "Fixed device latency: ..." in the trace log).

According to [this page](https://docs.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio), starting with Windows 10 the default audio engine latency is 1.3ms, plus a 10ms default period size which will get written to the buffer for the hardware. So adding that all up, there should be about 51.3ms to 71.3ms if there's no other hidden latency anywhere. By changing OpenAL Soft's period size and period count properties, It could be reduced to a period size of 10ms and a 20ms buffer, which would make OpenAL Soft average 15ms, making the latency from OpenAL to output about 21.3ms to 31.3ms. Although this will have a higher risk of underruns.

Before Windows 10, there's an additional 11ms for floating point sample streams and 5ms for integer sample streams. APOs may add additional latency, but there's no information about if there's any used normally.

> Alternatively, someone even modified OpenAL Soft to use WASAPI in exclusive mode, which I've tested and confirmed it does make a difference (tho I'm yet to measure it), so I forked it here https://github.com/ThreeDeeJay/openal-soft-WASAPI-exclusive/commit/9cd722fc9a80181cc9c86db9a0ec86728dafb7a3

Well, one apparent difference is it passes a bad period size to `IAudioClient::Initialize` (it passes the same size for the buffer and period size, using the buffer size, when the buffer size should be at least twice the period size), sets incorrect values for the OpenAL device's buffer and period size (sets the period size using the buffer size), and doesn't properly pace updates (whenever the mixer thread wakes up, it processes however many samples WASAPI says are available regardless if it's at the period size yet). It also seems to get the minimum period size before initialization and the buffer size after initialization, but does nothing with them. It's impossible to tell what the device is going to do with regards to buffering/latency.

**Comment by mirh** [comment]
I have always appreciated ASIO... if not any because I had a Xonar sound card, and even for my poor [realtek](https://www.baumannmusic.com/2021/the-official-asio-driver-for-realtek-hd-audio-dell-hp-lenovo-asus/) I had found a \*native\* driver anyway (also, I think they had made some multiclient driver?). 
But is there really much of a point in 2022 over a "normal" api like [IAudioClient3](https://docs.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio#windows-audio-session-api-wasapi) in exclusive mode? https://github.com/mumble-voip/mumble/issues/1604

I mean, putting aside that I don't think games are meaningfully hampered by this. Academically speaking, is it worth at least 1ms? Or is it just a relic of another epoch when the windows mixer was called *KMixer*?

p.s. as far as WDM-KS workarounds go.. I believe [FlexASIO](https://github.com/dechamps/FlexASIO) was the current champ

**Comment by ThreeDeeJay** [comment]
Never had luck getting less than 10ms with REAL.
![image](https://user-images.githubusercontent.com/71472458/171480984-a3e8ce00-e061-4970-89cd-2de976428455.png)
I should point out that I didn't revert back to the Microsoft drivers (which is optional anyway) cuz I wouldn't wanna lose 7.1/5.1 in both my internal/USB sound cards.

**Comment by Enokilis** [comment]
> Well, one apparent difference is it passes a bad period size to `IAudioClient::Initialize` (it passes the same size for the buffer and period size, using the buffer size, when the buffer size should be at least twice the period size), sets incorrect values for the OpenAL device's buffer and period size (sets the period size using the buffer size), and doesn't properly pace updates (whenever the mixer thread wakes up, it processes however many samples WASAPI says are available regardless if it's at the period size yet). It also seems to get the minimum period size before initialization and the buffer size after initialization, but does nothing with them. It's impossible to tell what the device is going to do with regards to buffering/latency.

I'm the one who made the modifications a long time ago. It was just a quick hack as a proof of concept, and wasn't really meant to be shared, no pun intended.

While not a controlled experiment, I used Wireshark with USBcap to measure the delta between a mouse click and a response in the audio stream. The advantage of this approach is that the DAC's own latency is factored out, but it assumes Wireshark is precise enough to be useful. Using the lowest period size I could in shared mode, it tended towards 30 milliseconds and up, while in exclusive mode, it was typically around 20 milliseconds.
As mentioned before, this is very app-dependent, and some OpenAL game could easily create a delay close to three digits of milliseconds, so exclusive mode is hardly a panacea.

**Comment by ThreeDeeJay** [comment]
Following kcat's suggestions [here](https://github.com/kcat/openal-soft/issues/999#issuecomment-2133543362), I was able to [compile OpenAL Soft with ASIO output via PortAudio](https://github.com/ThreeDeeJay/openal-soft/actions/runs/10016927693):
[OpenALSoft+PortAudio+ASIO.zip](https://github.com/user-attachments/files/16318757/OpenALSoft%2BPortAudio%2BASIO.zip)
![sublime_text_45POugxw0M](https://github.com/user-attachments/assets/2e546ab6-3fd0-4012-a981-6304a6147850)
![image](https://github.com/user-attachments/assets/44af8139-381c-4906-b34f-daf54d0fa7b7)    ![sublime_text_7hZrPKb15p](https://github.com/user-attachments/assets/0bff706f-92ed-4b73-964a-b446d5655a25)
![na3tXhQ8DC](https://github.com/user-attachments/assets/3b6053e1-f72c-4f1b-8417-b0c1051b62c3)

However, I'm not sure how to force set buffer size to 64 samples (lowest my sound card can handle in native ASIO apps) for the lowest possible latency because even after setting `period_size=64` it keeps resetting to much higher values and the slider in ASIO4ALL gets ignored, so perhaps there's something that I'm missing? 🤔 
[alsoft_error.txt](https://github.com/user-attachments/files/16318763/alsoft_error.txt)

On a side note, adding [DSOAL+RightMark3DSound.zip](https://github.com/user-attachments/files/16318765/DSOAL%2BRightMark3DSound.zip) (specifically dsound.dll) [breaks EFXShow](https://i.imgur.com/rDpXI5g.png) for some reason, and [RightMark3DSound crashes too](https://i.imgur.com/5uT8MM8.png) with this build.

**Comment by kcat** [comment]
> However, I'm not sure how to force set buffer size to 64 samples (lowest my sound card can handle in native ASIO apps) for the lowest possible latency because even after setting `period_size=64` it keeps resetting to much higher values and the slider in ASIO4ALL gets ignored, so perhaps there's something that I'm missing? 🤔

Currently the way the PortAudio backend works is it opens and configures the output stream during `alcOpenDevice`, when the actual ALCdevice configuration isn't handled until `alcCreateContext` (and `alcResetDeviceSOFT`, etc), so the PortAudio stream gets configured with the default properties. A way to fix this would be to recreate the PortAudio stream in `PortPlayback::reset`, but that risks failing if the device doesn't like being reopened immediately after closing, or if there's trouble getting it working with a usable format, making the device unusable.

> On a side note, adding [DSOAL+RightMark3DSound.zip](https://github.com/user-attachments/files/16318765/DSOAL%2BRightMark3DSound.zip) (specifically dsound.dll) [breaks EFXShow](https://i.imgur.com/rDpXI5g.png) for some reason, and [RightMark3DSound crashes too](https://i.imgur.com/5uT8MM8.png) with this build.

Probably a dependency loop. `0xC0000142` is `STATUS_DLL_INIT_FAILED`, and since PortAudio can use DSound, having OpenAL Soft load PortAudio, which loads DSound/DSOAL, which loads OpenAL Soft, creates a circular loop which causes the DLL to fail initialization.

**Comment by mirh** [comment]
On a separate note, if you have a realtek you should try native asio instead of asio4all.

**Comment by ThreeDeeJay** [comment]
@kcat Also I tried using the [latest commit instead of latest stable version](https://github.com/ThreeDeeJay/openal-soft/commit/e32151aff56727b4dba5ef03aa6a7199fb1e6226) (from 2021) with no luck, and I found a [PR that implements ASIO messages](https://github.com/PortAudio/portaudio/pull/519) and [rebuilt the dll with it](https://github.com/ThreeDeeJay/openal-soft/commit/cb84ac20bcd709b6c7e3ea288e4ff529694a0046), but *Buffer Size Changed* via the ASIO4ALL control panel still gets ignored. e.g. here I set it to request 64 samples (which works with native ASIO apps, at least after a restart) but it wouldn't change from 1920 as the log below shows
![image](https://github.com/user-attachments/assets/ce5151fe-1196-411c-9e2e-ad332aca0e76)

> Probably a dependency loop. `0xC0000142` is `STATUS_DLL_INIT_FAILED`, and since PortAudio can use DSound, having OpenAL Soft load PortAudio, which loads DSound/DSOAL, which loads OpenAL Soft, creates a circular loop which causes the DLL to fail initialization.

I wonder if disabling DirectSound support from PortAudio would get around that issue.
It would be interesting to check whether apps/games using at least DirectSound would get low latency via this ASIO route 🤔 

> On a separate note, if you have a realtek you should try native asio instead of asio4all.

@mirh Sadly, my motherboard's onboard Realtek ALC1150 drivers don't include an ASIO driver, and I've tried the [Dell Realtek drivers](https://www.dell.com/support/home/en-us/drivers/driversdetails?driverid=7776f), even with [this installer](https://github.com/KeppySoftware/RealtekASIOInstaller), but I ran into the same issue, even at 44100hz which some have reported to work more reliably:
```
[ALSOFT] (II) Created device 02E32960, "OpenAL Soft on Realtek ASIO"
[ALSOFT] (II) Found option frequency = "44100"
[ALSOFT] (II) Found option period_size = "128"
[ALSOFT] (II) Found option stereo-encoding = "hrtf"
[ALSOFT] (II) ALC_MAX_AUXILIARY_SENDS = 2
[ALSOFT] (II) Pre-reset: Stereo, Float32, *44100hz, 128 / 384 buffer
[ALSOFT] (II) Reported stream latency: 0.002979 sec (143.000000 samples)
[ALSOFT] (WW) Failed to set 44100hz, got 48000hz instead
[ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 960 / 1920 buffer
[...]
[ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 960 / 1920 buffer
```
It also suffers from other bugs like [inputs/outputs randomly disappearing](https://images-ext-1.discordapp.net/external/8s40hwILNKRBYMtZs7IFELjCa6ArwVNfIPfOYhIQAu8/https/i.imgur.com/KzefIom.mp4), some apps reporting single, high buffer size like [888](https://i.imgur.com/FaUP0QC.png) or [960](https://github.com/user-attachments/assets/580efa12-1aeb-4bab-b3b6-4cd6e74da22e). Using [Creative's generic ASIO drivers on my X-Fi is even worse](https://github.com/user-attachments/assets/97d341de-4d0d-4aea-87c6-dd8fdbe13ca5) so I just use [ASIO4ALL](https://github.com/user-attachments/assets/f23a26b3-80ee-4455-a097-1e14f230db3c) which just works:tm: 99% of the time.

**Comment by ThreeDeeJay** [comment]
*I spy with my little eye 👀* https://github.com/kcat/openal-soft/commit/aafaf6c6669da366e1833d23f677ce81e4b7dda8
Good news, now it's reporting much lower buffer size, though not quite the lowest.
I specified `period_size=64` and `periods=2` but ASIO4ALL still refuses to go below `128` for some reason.
[alsoft_error.txt](https://github.com/user-attachments/files/16324898/alsoft_error.txt)
```
[ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 64 / 128 buffer
[ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 64 / 172 buffer
[ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 64 / 172 buffer
[ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 64 / 128 buffer
[ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 64 / 176 buffer
[ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 64 / 176 buffer
```
![EFX10ShowWin32_p9LQ83FZ0k](https://github.com/user-attachments/assets/3dacb448-2210-4297-81dc-68243df88b8f)
For reference, here's how it should look:
![image](https://github.com/user-attachments/assets/ebcb0447-fee9-4b6f-9b75-54b1bbbbf145)

**Comment by mirh** [comment]
Did you try with some older drivers? ALC1150 is at least a decade old (UAD in particular is pretty [delicate](https://github.com/pal1000/Realtek-UAD-generic/issues/72))
Or maybe with the hacked ones on TPU.

**Comment by kcat** [comment]
> Good news, now it's reporting much lower buffer size, though not quite the lowest. I specified `period_size=64` and `periods=2` but ASIO4ALL still refuses to go below `128` for some reason.
> [alsoft_error.txt](https://github.com/user-attachments/files/16324898/alsoft_error.txt)

It can't go less than 128 with an update size of 64. Some samples need to be playing while new samples are being generated, which is accomplished with double-buffering, and 64x2 = 128. Though it looks like it's not going lower than 176 (~3.6ms), which is x2.75. That could be a limit of PortAudio, to ensure there's enough time to call for more audio before underrunning, but OpenAL Soft is only asking for 128-sample latency for double-buffering, and is getting back 176.

**Comment by ThreeDeeJay** [comment]
> Did you try with some older drivers? ALC1150 is at least a decade old (UAD in particular is pretty [delicate](https://github.com/pal1000/Realtek-UAD-generic/issues/72)) Or maybe with the hacked ones on TPU.

@mirh Any idea if those drivers perform any differently than ASIO4ALL? 🤔
Seems a bit tedious and unsafe if it might also require disabling driver signature enforcement to install modified drivers.
I even had to revert drivers [R2.83](https://www.realtek.com/Download/List?cate_id=593&menu_id=298) released earlier this year because 7.1 surround configuration was missing so I went back to R2.82 from like 2017 lol


> It can't go less than 128 with an update size of 64. Some samples need to be playing while new samples are being generated, which is accomplished with double-buffering, and 64x2 = 128. Though it looks like it's not going lower than 176 (~3.6ms), which is x2.75. That could be a limit of PortAudio, to ensure there's enough time to call for more audio before underrunning, but OpenAL Soft is only asking for 128-sample latency for double-buffering, and is getting back 176.

@kcat I noticed there may be a pattern here:

- frequency=**44**100
	- period_size=64
		- \* 2 = 128 double buffer
			- \+ **44** = 172 actual buffer
				- `[ALSOFT] (II) Post-start: Stereo, Float32, 44100hz, 64 / 172 buffer`
- frequency=**48**000
	- period_size=96
		- \* 2 = 192 double buffer
			- \+ **48** = 240 actual buffer
				- `[ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 96 / 240 buffer`
- frequency=**96**000
	- period_size=64
		- \* 2 = 128 double buffer
			- \+ **96** = 224 actual buffer
				- `[ALSOFT] (II) Post-start: Stereo, Float32, 96000hz, 64 / 224 buffer`

So if my guess and math are right, ActualBuffer = (`period_size` * 2) + (frequency/1000), then given ActualBuffer = 64 and `frequency=48000`, `period_size` would need to be 8 but that's way below the [acceptable values](https://github.com/kcat/openal-soft/blob/aafaf6c6669da366e1833d23f677ce81e4b7dda8/alsoftrc.sample#L85).
<details>
  <summary>Math</summary>

(`period_size` * 2) + (frequency/1000) = 64
 ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞2 ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞ ͞   ͞ ͞ ͞2 ͞ 
`period_size` + [(48000/1000)/2] = 32
`period_size` + (48/2) = 32
`period_size` + 24 = 32
`period_size` = 32 - 24
`period_size` = 32 - 24
`period_size` = **8**

</details>

Alternatively, If I could set `periods=1`, I could just set `period_size=16` so that extra `+ 48` at 48000Hz adds up to 64, but `periods=1` [isn't acceptable](https://github.com/kcat/openal-soft/blob/aafaf6c6669da366e1833d23f677ce81e4b7dda8/alsoftrc.sample#L94) either.

So would it be feasible to lower those limits to compensate for that extra buffering that's added anyway?
I wonder if it'd increase CPU usage significantly tho, at least compared to native ASIO at 64 buffer.

Also worth noting that I'm still able to use 64 samples even at up to 192000hz (max supported in general) in other ASIO apps without a single crackle.
![image](https://github.com/user-attachments/assets/55f6ba7b-6647-4379-a4d3-a1eea0651cc3)

**Comment by mirh** [comment]
I don't have any idea, other than audio vendors not having needed ASIO in the first place if WDM-KS had been enough.
Anyhow, whatever it's just an audio driver. Even R2.79 has its admirers

**Comment by mirh** [comment]
There is some [shaky](https://github.com/PortAudio/portaudio/issues/385) report the new W10 low latency mode may get you 3ms latencies, but it's really freaking annoying how no competent developer can seem to independently get it to work and confirmed. 

p.s. as for the realtek asio driver, I found mixed opinions: [one](https://audiosex.pro/threads/realtek-asio-driver-v1-0-0-1.3071/) super positive, one [neutral](https://forum.cockos.com/showthread.php?t=205623) (old version sucks royally while new one is good, but pretty much the same of WASAPI) and [another](https://forum.pianoworld.com/ubbthreads.php/topics/2683364/realtek-asio-anyone-used-it.html) negative.

**Comment by ThreeDeeJay** [comment]
<5ms latency in shared mode sounds too good to be true, but then again so did the graphics equivalent  (fullscreen optimizations/flip model or whatever it's called) add it really turned out to be a decent middle middle ground between the performance and latency of exclusive fullscreen, without its inconveniences like not being able to draw regular windows on top of it or without alt-tabbing, so I wonder if this would be feasible here as well, to reduce inconvenience and extra setup for the end user 🤔

**Comment by dechamps** [comment]
> <5ms latency in shared mode sounds too good to be true

There's really no reason why that shouldn't be possible, but to me the main caveat is this "low latency shared mode" apparently requires explicit support from the audio driver. I don't know if typical drivers offer such support (hopefully at least the Microsoft USB Audio drivers and Realtek drivers do, otherwise that's a huge chunk of the market left unadressed). I've never really looked into this particular feature.

**Comment by kcat** [comment]
Commit 246d50d7b82c18f799e774918037f18428bda04c adds a config option for using WASAPI exclusive mode. It's experimental, I can't test it, and I don't know if Windows may have some quirky behavior with the reported period and buffer sizes, but it's there to try out.

**Comment by ThreeDeeJay** [comment]
@kcat Neat! It activates exclusive mode, but the audio's not quite right. Tested on Realtek ALC1150 (internal), Creative X-Fi Surround 5.1 (USB) and the PlayStation 5 DualSense controller (USB) because why not.
- It requires `sample-type=int16` to prevent crashing, like the [old fork](https://github.com/ThreeDeeJay/openal-soft/releases/tag/1.22.2-WASAPI-exclusive)
- If I use `period_size=144` (lowest value that doesn't crash, though my onboard Realtek requires 160), it sounds kinda robotic/low bitrate and slowed down. And if I use `period_size=65536`, playback speed and quality seems normal but but there's still frequent pops/crackles.
- I'm getting no audio on my onboard Realtek, which ironically is the only one that works with the [old fork](https://github.com/ThreeDeeJay/openal-soft/releases/tag/1.22.2-WASAPI-exclusive) where Creative X-Fi USB has no audio and DualSense crashes
- The system volume doesn't work on my Creative X-Fi USB (not uncommon when using WASAPI exclusive), but weirdly enough, it does work on my DualSense, which by the way [only supports quadraphonic configuration](https://github.com/user-attachments/assets/0eb31692-c1d3-4877-ac91-317f053b2024) (first 2 channels are the regular stereo for headphones/built-in speaker and the last 2 are for haptics/vibration)

Test files and logs: [OpenAL Soft + DSOAL - WASAPI Exclusive test.zip](https://github.com/user-attachments/files/18138011/OpenAL.Soft.%2B.DSOAL.-.WASAPI.Exclusive.test.zip)

**Comment by dechamps** [comment]
> The system volume doesn't work on my Creative X-Fi USB (not uncommon when using WASAPI exclusive), but weirdly enough, it does work on my DualSense

This is normal and expected if the Creative has hardware volume control, but the DualSense doesn't. In any case, there's nothing an application can do about it.

**Comment by kcat** [comment]
```
[ALSOFT] (II) Found option period_size = "65536"
[ALSOFT] (II) Found option periods = "1"
```
That is odd. I don't know what requesting such a high period size will do (which gets clamped to 8192). WASAPI seems to report back 10ms (480 samples) for the period size, but 8192 for the buffer size. It may be more helpful to see logs using more normal (or other low values that are expected to work) values. As well as logs using normal OpenAL apps, rather than DSOAL (if OpenAL Soft is having issues mishandling the buffer or period size, it could influence DSOAL behavior and cause extra issues).

**Comment by ThreeDeeJay** [comment]
Yup, I set it to 65536 myself to see how high I'd need to go to stop popping/crackling but yeah I noticed after one point it stopped improving, so here's another test with `period_size=960` (default for 48KHz) using OpenAL SDK's EFX10ShowWin32:
[EFX10Show - WASAPI Exclusive test.zip](https://github.com/user-attachments/files/18138651/OpenAL.Soft.-.WASAPI.Exclusive.test.zip)

And here's another one with [OpenAL Minerva](https://sourceforge.net/projects/openal-minerva/files/latest/download):
[OpenAL-Minerva - WASAPI Exclusive test.zip](https://github.com/user-attachments/files/18138764/OpenAL-Minerva.-.WASAPI.Exclusive.test.zip)

<details>
  <summary>Minerva Tests</summary>

  ```
This application is currently running as x86.
Renderer: OpenAL Soft
Total XRam: 64MB
Free XRam: 64MB
Free XRam after loading files: 64MB

Testing 2D Mixing.
Playing stereo source of 22kHz and 16-bit.
Mixing with mono source of 44kHz and 16-bit.
Mixing with stereo source of 22kHz and 16-bit.
Mixing with mono source of 11kHz and 16-bit.
Mixing with mono source of 44kHz and 8-bit.
Mixing with mono source of 44kHz and 16-bit.

Testing Sample Rate Conversion.
Loading mono sample of 11kHz and 16-bit.
Initial playback at normal rate, followed by...
sample rate conv. with factor range of <0.5 - 3.5>.
Loading mono sample of 22kHz and 16-bit.
Initial playback at normal rate, followed by...
sample rate conv. with factor range of <0.5 - 3.5>.

Testing Distance...
for mono file of format 44kHz and 16-bit
-Doppler effect turned off
   with MIN=1.0m and MAX=35.0m

   with MIN=1.0m and MAX=infinity

Testing Doppler...
 for mono source of 22kHz and 16-bit
  with speed of 36 km/h
  with speed of 108km/h
  with speed of 36km/h
  with speed of 108km/h
  with speed of 90km/h and doppler factor of 1.0
  with speed of 90km/h and doppler factor of 2.0
  with speed of 90km/h and doppler factor of 4.0

Testing Positioning with single source...
 mono of 44kHz and 16-bit
 -Doppler effect turned off
Source first rotates around the head twice, with 2m radius...
Then rises vertically to the front left of the listener
 and falls vertically to the rear right of the listener

Testing multi source positioning...
Red source format: 44kHz 16-bit
Blue source format: 22kHz 8-bit
Green source format: 44kHz 16-bit
Pink source format: 44kHz 8-bit

Testing Radiation...
Inner Cone angle = 90 degrees
Outer Cone angle = 270 degrees
Source position = (0, 0, -2)
Source position = (0, 0, 2)

Testing Latency...

Testing EFX Echo effect...
 with mono source of 22kHz and 8-bit...
 first without effects
 now with effects
 and with mono source of 44kHz and 16-bit

Testing EFX Reverb effect...
 with mono source of 22kHz and 8-bit...
 first without effects
 now with effects
 and with mono source of 44kHz and 16-bit

Testing EFX EAX Reverb effect...
 with mono source of 22kHz and 8-bit...
 first without effects
 now with effects
 now one rotation of the source at a distance of 4m
 now one rotation of the source and the effects at a distance of 4m

Testing EFX Occlusion effect...
 with mono source of 22kHz and 8-bit...
 first fully occluded
 now disabling occlusion
 and enabling it again

Testing EFX Exclusion effect...
 with mono source of 22kHz and 8-bit...
 first showing the transition with reverb that is transmited through walls
 now showing the transition with reverb that can only go through the aperture

Testing EFX Obstruction effect...
 with mono source of 22kHz and 8-bit...

All tests ended.

  ```
  
</details>

Results were similar, but worse:
- Audio slow down more pronounced
- More frequent pops/crackling and longer (especially with the DualSense where it's more like audio getting intermittently getting cut off instead of shorter crackling like with the Creative card)
- Still no audio from my Realtek device
- Increasing `periods` doesn't seem to help
- Not using `sample-type=int16` still crashes

On a side note, I don't remember the [log getting flooded with errors when moving the sound emitter position in EFXShow](https://github.com/user-attachments/assets/4ea6672d-de54-4f4e-8e34-d57faaa68192). 
If anyone else can confirm on their end I can open a new issue 👀

**Comment by ThreeDeeJay** [comment]
New test with the build from @LAGonauta's PR https://github.com/kcat/openal-soft/pull/1084

Now I'm getting audio on my Realtek onboard, and even without any slowdowns or pops, like the old fork, though the lowest period size I can reach without crashing is 160@48000hz on my Realtek. I remember being able to run 96@44100hz in the old fork. My DualSense and X-Fi USB devices can get down to 144@48000hz, but they still suffer from the same severe audio speed/quality issues even at high period size. IIRC all my devices can get down to 64 samples when using ASIO, so I wonder if WE is just not able to reach such low buffer/period sizes 🤔 

Anyway, I included the logs here:
[LAGonautaPR - OpenAL Soft - WASAPI Exclusive test.zip](https://github.com/user-attachments/files/18144288/LAGonautaPR.-.OpenAL.Soft.-.WASAPI.Exclusive.test.zip)

**Comment by kcat** [comment]
The Realtek crash with 144 period size seems to be due to `IAudioClient::Initialize` failing with `AUDCLNT_E_BUFFER_SIZE_NOT_ALIGNED`. This isn't handled yet, so the reset fails. I haven't been able to dig too deep into it, but for some reason, if resetting fails, a crash happens at some later point, even though it should be able to clean up fine.

The DualSense crash is because it's an unsupported format, `IAudioClient::IsFormatSupported` returns false and it tries to use the default mixing format as a fallback, which it doesn't like since it's a float format and initialization fails. The X-Fi crash is a little weird... `IAudioClient::Initialize` fails with `AUDCLNT_E_UNSUPPORTED_FORMAT`, but `IAudioClient::IsFormatSupported` succeeded just before, so I don't know what it doesn't like.

**Comment by ThreeDeeJay** [comment]
@kcat It might be worth pointing out that in addition to Quad config only, the DualSense [only supports 16-bit 48000hz](https://github.com/user-attachments/assets/c25a0802-4807-4998-a609-ca965c734788)
As for the Creative card, I noticed the drivers didn't actually install for some reason (guess Windows 11 hates PAX drivers?) so I just installed the official version and now everything's working fine: [at 44100Hz I can even get down to `period_size=133`](https://www.diffchecker.com/1vPy8Pql/) and in high quality (no longer [bit crushed](https://youtu.be/QLyhfakB-QY)). Sadly I can't find any newer/official Sony drivers for the [DualSense](https://devicehunt.com/view/type/usb/vendor/054C/device/0CE6), only [the generic Windows one](https://github.com/user-attachments/assets/be44d4cf-14a0-4341-be3b-f2e61059c9c9) .

**Comment by ThreeDeeJay** [comment]
> https://github.com/kcat/openal-soft/pull/1084#issuecomment-2568645246
Possibly the Realtek device can't go lower than 160 (~3.333ms), while the Creative card can go a bit lower. If initializing fails with `AUDCLNT_E_BUFFER_SIZE_NOT_ALIGNED`, it gets the next supported period size, which would be 160. Values less than 133 are probably triggering a different initialization error (`AUDCLNT_E_BUFFER_SIZE_ERROR` maybe?), which causes a failure and the delayed crash.

@kcat That makes sense.
I found a neat program called [FreePiano](https://github.com/LJFloor/freepiano) to test a MIDI/computer keyboard with multiple output APIs and even reports latency which seems about right though I'm not sure how accurate it is, and the sample-rate/bit-depth sometimes change unpredictably.
- WASAPI shared reports 10ms
![image](https://github.com/user-attachments/assets/1d02dbf1-583e-41d9-bd0c-01a83be0acbb)
- Same for DirectSound, which makes sense since it uses WASAPI shared (Exclusive option makes no difference)
![image](https://github.com/user-attachments/assets/8c43113e-4ac4-49f6-ba6c-4337d99720e7)
- With DSOAL using WASAPI exclusive build/cfg, the improvement in latency is subtle but noticeable, even if it's just 6.4ms. I wonder if allowing a single period in OpenAL Soft would be feasible/beneficial 🤔 
![image](https://github.com/user-attachments/assets/980ced41-877d-4755-9ffa-b0235fed5f68)
- WASAPI exclusive reports 3.3ms, which matches 160 buffer like you mentioned so I guess that's just the lowest the device can go in WASAPI exclusive 🤔 
![image](https://github.com/user-attachments/assets/97420713-b31a-47cb-97ac-1d11246f43f5)
- ASIO is still able to go even lower as usual
![explorer_pr3x1tSF6A](https://github.com/user-attachments/assets/28839972-0f0e-4840-ab93-c0aca8d89815)
It's *almost* indistinguishable from DSOAL but tbh I doubt ~2ms(?) would be worth the hassle of setting up ASIO (not to mention implementing it in OpenAL Soft).

Btw, apparently `sample-type=int16` isn't needed anymore, so I guess the only thing left to do before we can close this issue is to prevent no audio/crash when setting periods too low (perhaps `exclusive-mode=true` should automatically find the lowest supported `period_size` if it's not specified already, to simplify setup) and maybe get some more testers so it's considered stable and this can be closed. 👀👌

**Comment by kcat** [comment]
> With DSOAL using WASAPI exclusive build/cfg, the improvement in latency is subtle but noticeable, even if it's just 6.4ms. I wonder if allowing a single period in OpenAL Soft would be feasible/beneficial 🤔

I'm curious how it calculates the latency with DSound. The only place DSOAL reads OpenAL update rate is `DSound8OAL::notifyThread`, which is just checking source offsets/state to trigger any event handles the app sets for DSound buffers, and doesn't report it to the app. I don't see anything [in the source](https://github.com/LJFloor/freepiano/blob/master/src/output_dsound.cpp) where it calculates the latency either.

> Btw, apparently `sample-type=int16` isn't needed anymore, so I guess the only thing left to do before we can close this issue is to prevent no audio/crash when setting periods too low (perhaps `exclusive-mode=true` should automatically find the lowest supported `period_size` if it's not specified already, to simplify setup) and maybe get some more testers so it's considered stable and this can be closed. 👀👌

The period size should now get clamped to the lowest supported value, with commit ee61ef398c837f06f99502c45bf1964964a46ef2. I don't know what's causing the crash when there's an error preventing the device from being reset, since it doesn't seem to be in OpenAL Soft itself I don't know where to look. More testing from more people would be nice in either case.

**Comment by ThreeDeeJay** [comment]
@kcat Clamping works as expected 👌
https://www.diffchecker.com/otDkbAIh/

By the way, would it be a good idea for the period size to get automatically clamped to the lowest supported value by default on WASAPI exclusive when the period size isn't set? Lowest possible latency is probably the main reason why people would use it (besides bypassing system effects).
And on that note, could `periods=1` ever work now that we have exclusive mode? or is there a fundamental reason/hardware limitation why 2 would still be the  [lowest acceptable value](https://github.com/kcat/openal-soft/blob/ee61ef398c837f06f99502c45bf1964964a46ef2/alsoftrc.sample#L94)?

By the way, I noticed something *really* odd:
```ini
[General]
period_size=1
periods=1

[wasapi]
exclusive-mode=true
```
When using just that config (removed stuff setting frequency/stereo/headphone/HRTF), I noticed that setting the speaker config to 7.1 or Quadraphonic in the Windows sound panel clamps to an even lower value (144) compared to 160 in Stereo and 5.1.  🤔 
https://www.diffchecker.com/7lbpLQFO/
https://www.diffchecker.com/NBJQLmxI/
Same with `frequency=44100`
https://www.diffchecker.com/8dqoFpnm/

**Comment by kcat** [comment]
> By the way, would it be a good idea for the period size to get automatically clamped to the lowest supported value by default on WASAPI exclusive when the period size isn't set? Lowest possible latency is probably the main reason why people would use it (besides bypassing system effects).

I'd be wary for performance concerns. It's fine if the user wants to play around and find what works for their system, but the defaults should reliably work. Exclusive mode will automatically drop to 2 periods (WASAPI itself forces this for exclusive mode), but the period size probably shouldn't act differently.

> And on that note, could `periods=1` ever work now that we have exclusive mode? or is there a fundamental reason/hardware limitation why 2 would still be the [lowest acceptable value](https://github.com/kcat/openal-soft/blob/ee61ef398c837f06f99502c45bf1964964a46ef2/alsoftrc.sample#L94)?

It's a limitation of how the system works. The device needs to have samples to play while more samples are being generated. Just like how with graphics, you have one image buffer that's being physically shown on the monitor, while rendering the next image on a separate unseen image buffer. It's a classic double-buffer setup; write to one buffer while the other buffer is being presented to the user, so that when the other buffer is done being presented to the user, the next one is ready and gets swapped in seamlessly. Unless the code is poking physical hardware bits at precisely timed intervals for the DAC (which is typically at the hardware or firmware level, modern OSs don't give the necessary timing and performance guarantees to do it with software), you need a second buffer/period to prepare samples in ahead of time.

Regardless, in the case of WASAPI exclusive mode, the client (OpenAL Soft) has no say in the number of periods, it's always 2.

> When using just that config (removed stuff setting frequency/stereo/headphone/HRTF), I noticed that setting the speaker config to 7.1 or Quadraphonic in the Windows sound panel clamps to an even lower value (144) compared to 160 in Stereo and 5.1. 🤔

Interesting, maybe some restrictions on the period sizes the hardware can use (for the number of bytes rather than the number of sample frames). Each period invokes a hardware interrupt, so there's probably limits on how often the interrupt can be fired while processing the hardware buffer.

**Comment by ThreeDeeJay** [comment]
> I'd be wary for performance concerns. It's fine if the user wants to play around and find what works for their system, but the defaults should reliably work. Exclusive mode will automatically drop to 2 periods (WASAPI itself forces this for exclusive mode), but the period size probably shouldn't act differently.

@kcat Then I guess we could just direct anyone who wants the lowest possible latency to the configs added in https://github.com/kcat/openal-soft/pull/1094 👌 

> It's a limitation of how the system works. The device needs to have samples to play while more samples are being generated. Just like how with graphics, you have one image buffer that's being physically shown on the monitor, while rendering the next image on a separate unseen image buffer. It's a classic double-buffer setup; write to one buffer while the other buffer is being presented to the user, so that when the other buffer is done being presented to the user, the next one is ready and gets swapped in seamlessly. Unless the code is poking physical hardware bits at precisely timed intervals for the DAC (which is typically at the hardware or firmware level, modern OSs don't give the necessary timing and performance guarantees to do it with software), you need a second buffer/period to prepare samples in ahead of time.
> 
> Regardless, in the case of WASAPI exclusive mode, the client (OpenAL Soft) has no say in the number of periods, it's always 2.

Does that mean ASIO actually uses 128 sample buffer when it says it's using 64, or does it just have more precise hardware timings that allow single period which WASAPI lacks even in exclusive mode? 
@dechamps Might know more about it 👀

**Comment by dechamps** [comment]
> Does that mean ASIO actually uses 128 sample buffer when it says it's using 64

Yes. ASIO uses a double buffering system. When you ask ASIO for a buffer size of N, that refers to the size of a single buffer, and the total buffer size is N*2. This is indeed a bit confusing - ideally ASIO should have called this number something like "periodicity" (which is the term other APIs like WASAPI use), i.e. the application gets a callback every N samples.

**Comment by ThreeDeeJay** [comment]
Alright, I've been using WASAPI exclusive pretty much exclusively (pun intended) and so far it's pretty stable in most devices I've tried (Realtek onboard, Creative USB, Acer monitor HDMI+headphone jack) so I guess I can go ahead and close this issue.

Though my DualSense still fails to start even with the default period size, and I just found out `channels=quad` works but then I lose HRTF (and rear channels too, since channels 3 and 4 are used for controller haptics so I literally only *feel* them lol)

Which begs the question: @kcat would it be feasible to fill/repeat channels to adapt to stubborn devices that lack regular stereo configuration?
It could also allow having more than 1 pair of headphones at a time by repurposing otherwise unused surround ports, especially if OpenAL Soft eventually gets support for [AutoEQ](https://github.com/jaakkopasanen/AutoEq/tree/master/results) like HeSuVi/EqualizerAPO, which get bypassed in exclusive mode.
It'd also allow the DualSense to have haptics, and at least in my case, I could use a different headphone port because the front/green port is *particularly* prone to GPU coil whine.

**Comment by kcat** [comment]
> Which begs the question: @kcat would it be feasible to fill/repeat channels to adapt to stubborn devices that lack regular stereo configuration?

OpenAL Soft should be able to output stereo even if the device is some other configuration, at least if stereo is explicitly requested. Though exclusive mode throws a wrench into that. Normally if you request stereo while the device can't use it, it would fallback to what it can use and check if the requested format can fit into the used format, in which case it'd mix the requested format and fill the extra channels with silence. But exclusive mode won't give a fallback format, you either get what you asked for, or it fails (technically I implement a manual fallback to 16-bit samples, but that's not the hardware device saying it needs that specifically, it's just a blind guess). So there's currently no way with exclusive mode to request stereo, get quad, then output stereo as requested with the rear channels left silent. It would need extra code to manually test all the various possible formats the device can use.

**Comment by mirh** [comment]
> It's almost indistinguishable from DSOAL but tbh I doubt ~2ms(?) would be worth the hassle of setting up ASIO (not to mention implementing it in OpenAL Soft).

It already is through portaudio as you tested?

> first 2 channels are the regular stereo for headphones/built-in speaker and the last 2 are for haptics/vibration

Wait a minute. Doesn't this mean that you get crappier audio this way? Like.. the front speakers in a non-stereo configuration don't sound the same if they were setup alone, is it? If a mixer knows you have more possible sound "granularity" it will remove certain effects (or volume) from them to be put in the others.

**Comment by kcat** [comment]
> Wait a minute. Doesn't this mean that you get crappier audio this way? Like.. the front speakers in a non-stereo configuration don't sound the same if they were setup alone, is it? If a mixer knows you have more possible sound "granularity" it will remove certain effects (or volume) from them to be put in the others.

It's a PlayStation (4? 5?) controller, I don't think it's meant to be used as a normal audio device on PCs. It would be specifically coded for in the game and/or console OS where it's behavior is known, accessing it as a four-channel device and fill the first two channels with plain stereo/headphone audio, and the second two channels with haptic strength.

**Comment by mirh** [comment]
Yes, of course (and actually it's [indeed](https://old.reddit.com/r/Dualsense/comments/197v8a5/dualsense_on_pc_uses_quadchannel_any_way_to_make/) as bad as I thought and you loose channels 3 and 4).
But be it somehow filling the two extra channels with silence or mirroring the front channels again (or maybe just treating them as LFE?) *then* I feel like it's kinda awful that the workarounds should go here rather than in the driver.

**Comment by ThreeDeeJay** [comment]
> It already is through portaudio as you tested?

@mirh Yep, but it's a [PITA to set up](https://github.com/kcat/openal-soft/issues/682#issuecomment-2240910728) (requires[ compiling with Steinberg SDK](https://github.com/kcat/openal-soft/compare/master...ThreeDeeJay:openal-soft:refs/heads/asio#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR123-R126) which we can't even redistribute because of licensing shenanigans, and even when it works, it sets the buffer way to high and I never figured out how to change it, and [Jack isn't any better](https://github.com/kcat/openal-soft/pull/1033).
But I was particularly referring to [OtherIO](https://github.com/kcat/openal-soft/commit/097035d21c510196bd0a241b239b70ef237efcd0) (OpenAL Soft's native ASIO backend) which AFAIK isn't fully functional atm and even if it was, most users would also need to install and configure ASIO4ALL or their native sound card ASIO, whereas WASAPI exclusive just needs a [4-liner INI edit](https://raw.githubusercontent.com/kcat/openal-soft/refs/heads/master/configs/WASAPI%20Exclusive/alsoft.ini).

> Wait a minute. Doesn't this mean that you get crappier audio this way? Like.. the front speakers in a non-stereo configuration don't sound the same if they were setup alone, is it? If a mixer knows you have more possible sound "granularity" it will remove certain effects (or volume) from them to be put in the others.

I'm not sure, though I do wonder if audio's any more compressed than if it was just stereo, since it's meant to be transmitted via Bluetooth (with a [custom protocol](https://github.com/streetpea/chiaki-ng/issues/271#issuecomment-1987943991)) but perhaps the [USB audio interface](https://github.com/Ryochan7/DS4Windows/issues/2738#issuecomment-1451244735) is still lossless?

> It's a PlayStation (4? 5?) controller, I don't think it's meant to be used as a normal audio device on PCs. It would be specifically coded for in the game and/or console OS where it's behavior is known, accessing it as a four-channel device and fill the first two channels with plain stereo/headphone audio, and the second two channels with haptic strength.

@kcat I'm not sure what you mean by normal audio device, but the [DualSense (PS5 controller) presents itself as a playback device](https://github.com/user-attachments/assets/df0d5980-0233-48e7-a742-83b137d4e43d), so any app/game can output to it, but yeah, haptics would require "surround" support. There's a [handful of games that do it natively](https://www.pcgamingwiki.com/wiki/List_of_games_that_support_Dualsense_haptic_feedback), though there are ways to repeat the system stereo audio into haptics like [DSX](https://store.steampowered.com/app/1812620/DSX/) and [DSY](https://github.com/WujekFoliarz/DualSenseY-v2) but HeSuVi is my favorite method because it allows [virtualizing up to 4.0 without a virtual audio device](https://github.com/user-attachments/assets/82f8cc88-16a6-4022-a316-f265de8c9ef0), and we can [EQ haptics](https://github.com/user-attachments/assets/0dc4cd7c-6dfe-4dd4-a689-30d784a09b32) with a [low-pass filter](https://github.com/user-attachments/files/18591278/DualSense.Haptics.txt) which supposedly improves it.

But unfortunately, none of those methods work with WASAPI exclusive (or ASIO)
So it would be great if OpenAL Soft could optionally repeat stereo onto surround pairs (or perhaps even reuse EqualizerAPO's config?) for multiple simultaneous headphones or haptics, and port-specific EQ but I suspect that's way outside of OpenAL Soft's scope 🥲

**Comment by mirh** [comment]
> Yep, but it's a https://github.com/kcat/openal-soft/issues/682#issuecomment-2240910728 (requires[ compiling with Steinberg SDK](https://github.com/kcat/openal-soft/compare/master...ThreeDeeJay:openal-soft:refs/heads/asio#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR123-R126) which we can't even redistribute because of licensing shenanigans

Yeah, LGPL wouldn't like it (even though I suppose that with over 95% of the code being written by a single person, that could even be somewhat abridged.. I believe jamulus does a similar shrug off). 
But even more if at the end of the day you aren't even native but using asio4all, why can't you use WDM/KS?

> But I was particularly referring to [OtherIO](https://github.com/kcat/openal-soft/commit/097035d21c510196bd0a241b239b70ef237efcd0) (OpenAL Soft's native ASIO backend)

Uh, damn. I was going to link [this](https://github.com/rleathart/SimpleASIO) and [this](https://github.com/s13n/cwASIO), but I'm very impressed (... even though it really seems overly lovely given what the main focus of oal-soft should be)

> and [Jack isn't any better](https://github.com/kcat/openal-soft/pull/1033).

I'm kinda having a bad time trying to figure out how that stacks against portaudio.

> install and configure ASIO4ALL or their native sound card ASIO, whereas WASAPI exclusive just needs a [4-liner INI edit](https://raw.githubusercontent.com/kcat/openal-soft/refs/heads/master/configs/WASAPI%20Exclusive/alsoft.ini).

FlexASIO and KoordASIO seem way more based.

> but perhaps the USB audio interface is still lossless?

?? It's not a matter of audio resolution, but mixing?
To be honest I cannot even understand why they went this way. Like, couldn't they just have two channels and recycle those? 
Unless that "quadraphonic" isn't KSAUDIO_SPEAKER_QUAD [but](https://en.wikipedia.org/wiki/Surround_sound#Standard_speaker_channels) KSAUDIO_SPEAKER_SURROUND and everything else (including the speaker setup graphics) is wrong.. but idk it seems so unlikely.

> HeSuVi is my favorite method because it allows [virtualizing up to 4.0 without a virtual audio device](https://github.com/user-attachments/assets/82f8cc88-16a6-4022-a316-f265de8c9ef0), and we can [EQ haptics](https://github.com/user-attachments/assets/0dc4cd7c-6dfe-4dd4-a689-30d784a09b32) with a [low-pass filter](https://github.com/user-attachments/files/18591278/DualSense.Haptics.txt) which supposedly improves it.

That's actually more surprisingly neat than I was expecting (honestly everybody else seems kinda grasping at dirt in comparison).. but as far as games are concerned you want* to let them work assuming just two channels? Since that's the number that you are eventually afforded through the controller to your ears.
Could HeSuVi do the mirroring? And if so then you could just have the low pass filter act as some kind of "vibration intensity slider".

*once upon a time I remember there was also the school of thought (my Xonar def works like that) that wanted to set your sound card on the most channels it supported for maximum "directional information", only for it to then downmix to your real number in-hardware... but for the sake of simplicity I'll put aside this generic dilemma.

> But unfortunately, none of those methods work with WASAPI exclusive (or ASIO)

Is low latency really that important for using buzz effects?
Putting aside that we aren't even IAudioClient3 ready yet AFAIU.

> way outside of OpenAL Soft's scope 🥲

Honestly, it feels way outside \*anything\* scope except a proper driver.

**Comment by ThreeDeeJay** [comment]
> but as far as games are concerned you want* to let them work assuming just two channels? Since that's the number that you are eventually afforded through the controller to your ears.

@mirh I'm not sure I follow, but we can easily switch between virtual 4.0 and stereo downmix (no HRTF), regardless of whether the app/game outputs stereo or 4.0/surround.

> Could HeSuVi do the mirroring? And if so then you could just have the low pass filter act as some kind of "vibration intensity slider".

It already does. It repeats the front channels into the side, rear and C/LFE pairs/ports so that we can have multiple headphones at the same time, but it just so happens to allow being repurposed for haptics.

> *once upon a time I remember there was also the school of thought (my Xonar def works like that) that wanted to set your sound card on the most channels it supported for maximum "directional information", only for it to then downmix to your real number in-hardware... but for the sake of simplicity I'll put aside this generic dilemma.

That reminds me, my Creative X-Fi allows "desyncing" the speaker configuration from Windows'. So it could also virtualize non-DS3D/OAL surround if the config in the Creative control panel is set to surround and CMSS-3D is enabled.

> Unless that "quadraphonic" isn't KSAUDIO_SPEAKER_QUAD [but](https://en.wikipedia.org/wiki/Surround_sound#Standard_speaker_channels) KSAUDIO_SPEAKER_SURROUND

Is there a way to check?
and I think the closest configuration would've been 2.2 (L+R LFE) but that's not a thing AFAIK

> Is low latency really that important for using buzz effects?

Generally not so much, but [it's a big plus](https://github.com/sicsix/MechShaker?tab=readme-ov-file#features), especially when haptics are supposed to be synced to the audio. Like, when the DualSense isn't the default playback device, [DSX](https://store.steampowered.com/app/1812620/DSX/) can mirror the other device's audio into the DualSense, but the latency is *really* jarring. But when the DualSense is the default playback device, the delay is negligible, even with HeSuVi (as long as a virtual audio device isn't used) so my concern was more that ASIO/WASAPI exclusive prevent HeSuVi from mirroring channel pairs at all because the system APO gets bypassed.

**Comment by mirh** [comment]
> So it could also virtualize non-DS3D/OAL surround if the config in the Creative control panel is set to surround and CMSS-3D is enabled.

Mmh, right. This trick makes sense when "you don't trust the mix" (so perhaps overwatch with atmos for headphones would be better to "keep real", while software directsound or fmod is better to be set to 7.1 if you have at least a realtek/turtlebeach/laptop that can do DTS/dolby headphone).

> Is there a way to check?

200% sure of it.. but idk.
With the most dumb search I found [this](https://github.com/davidebeatrici/WASAPI-GetMixFormat) FWIW.

> mirror the other device's audio into the DualSense, but the latency is really jarring. But when the DualSense is the default playback device

Mmmhh... Now that I think to it, I guess that fairly enough mirroring \*has to\* always incur some sort of latency (and idk how short you could make the WASAPI loopback buffer?).
Meanwhile directly outputting the same sound twice at the source would have none.

**Comment by MyGoodBear** [comment]
How can I redirect the Audio Data stream from all modern games to ASIO or at least to WASAPI Exclusive?

There is a problem, which is that the audio input/output quality via WASAPI is poor - this is due to a large delay and the passage of the signal through a lot of bad-sounding filters.

How do I send audio data from the game via ASIO, or at least via the WASAPI Exclusive shortcut?

We are not talking about OpenAL games, but about all modern games with WASAPI (Windows Mixer) support.

**Comment by dechamps** [comment]
> In the meantime, you could try installing EqualizerAPO and in the Configurator window, check Troubleshooting options and uncheck Use original APO. Then set your device's sample rate to match the game's and add a -0.1 pre-amp effect in EQAPO. That supposedly results in bit-perfect audio by disabling unwanted audio effects or processing like resampling and limiter, though it won't help with latency and I haven't actually noticed a difference myself and can't find the link to the forum where I saw it, but maybe it'd be helpful for you.

The shared Windows audio pipeline has been [independently verified to be bit-perfect](https://www.audiosciencereview.com/forum/index.php?threads/ending-the-windows-audio-quality-debate.19438/page-24#post-1194466) if all the following conditions are met:

- Device set to 24-bit output (Windows will dither if the output is 16-bit)
- Device output sample rate configured to match the original (otherwise Windows will automatically convert)
- All [system effect APOs](https://github.com/dechamps/APO) disabled (configuring Equalizer APO with no filters and "use original APO" unchecked is indeed one possible way to do this)
- [CAudioLimiter forcibly removed](https://www.audiosciencereview.com/forum/index.php?threads/ending-the-windows-audio-quality-debate.19438/page-24#post-1194253) through a registry hack

**Comment by MyGoodBear** [comment]
Crystal Mixer, по сути, преобразует вызовы OpenAL в ASIO.

Что мешает вам принимать вызовы из игр, поступающих в WASAPI, и перенаправлять их из ASIO?

Точно так же, как Creative ALchemy передает вызовы DirectSound в OpenAL.

**Comment by MyGoodBear** [comment]
"EqualizerAPO"

Я никогда не пользовался и никогда не буду пользоваться подобным хламом.

Мне нужен точный звук, а в WASAPI хлама более чем достаточно.

**Comment by MyGoodBear** [comment]
The exact sound is available only via ASIO, OpenAL hardware, and BitStream to an external decoder.

None of this is implemented in modern "AAA" games.

**Comment by MyGoodBear** [comment]
We need to solve this problem with games somehow.

**Comment by MyGoodBear** [comment]
That is, it is necessary to intercept those calls that come from the game on WASAPI and relay them to ASIO or, at the very least, hardware Openal AL (not all cards support it).

**Comment by dechamps** [comment]
> Don't mislead people.

I am not misleading anyone. I have verified this myself using a software loopback driver (VAC), and other people also confirmed this using real hardware - most notably an RME DAC that has a "bit-perfect verification" feature built-in using special test files. There are other ways this can be further verified, for example by recording a digital output such as S/PDIF and then checking the digital recording against the original stream.

You may want to properly research this stuff instead of simply rejecting anything you don't understand. For example your description of Equalizer APO as "trash" reveals a clear lack of knowledge about what Equalizer APO does and how it fits into the wider Windows audio pipeline.

I am not going to entertain this discussion further because I honestly don't think it is leading anywhere. You are clearly set in your ways and it seems unlikely anyone would be able to convince you otherwise.

**Comment by MyGoodBear** [comment]
"Device set to 24-bit output (Windows will dither if the output is 16-bit)"

"Device output sample rate configured to match the original (otherwise Windows will automatically convert)"

That's the whole point! Do you understand what you're writing?

Since the Windows mixer does not automatically adjust the card's device for depth (bit depth) and frequency, while still having a bunch of filters in the audio path, it turns out that I need to determine the sampling rate and bit depth of game samples in each game by ear.
Changing the frequency and bit rate for each game by ear through the Windows mixer, due to the fact that there is no automatic data machine, but everything is converted LOSSILY to the reference frequency and bit rate of the Windows mixer - this is the wild inconvenience of using and losing useful information due to the lack of automatic adjustment of the sound card device to the parameters of the samples in the game.

It's in ASIO and BitStream, and it's not in WASAPI.

**Comment by MyGoodBear** [comment]
"I have verified this myself using a software loopback driver (VAC), and other people also confirmed this using real hardware - most notably an RME DAC that has a "bit-perfect verification" feature built-in using special test files. There are other ways this can be further verified, for example by recording a digital output such as S/PDIF and then checking the digital recording against the original stream."


If the methods you use show that there is no difference between ASIO and WASAPI, then you are using incorrect information verification methods.

**Comment by MyGoodBear** [comment]
"You may want to properly research this stuff instead of simply rejecting anything you don't understand. For example your description of Equalizer APO as "trash" reveals a clear lack of knowledge about what Equalizer APO does and how it fits into the wider Windows audio pipeline."

It's not garbage - it's a hardware mixer: analog or DSP.
Software mixers are garbage.
The software mixer is a channel-by-channel cASIO I/O in 64-bit Float - with the ability to connect hardware mixers and hardware processing via VST - this is normal sound.
For example: Steinberg Nuendo or Cubase

The hardware compatibility of ASIO I/O is important.

All the other "ASIO mixers" made up of all sorts of homemade ones are garbage.

**Comment by MyGoodBear** [comment]
"I am not going to entertain this discussion further because I honestly don't think it is leading anywhere."

I see that you are the kind of person to whom everything sounds the same.

It was clear from your first message - subsequent messages only confirmed everything.

There is absolutely nothing to discuss here.

Do you know for whom everything always sounds and looks the same? - Think for yourself.

**Comment by ThreeDeeJay** [comment]
@dechamps Thanks for the update, that's exactly the thread I was referring to. 👌 

> Crystal Mixer essentially converts ASIO calls to OpenAL.

@MyGoodBear I don't think that's what it really does.
It uses a virtual audio device that forwards the system audio *via* ASIO to Crystal Mixer, which in turn outputs via OpenAL Soft which by default outputs to WASAPI on Windows (unless you add ASIO to OpenAL Soft as we discussed earlier in this thread, but even then I *really* doubt you'd be able to get lower latency than outputting to the device directly via WASAPI or gain other benefits like native ASIO's, and might even run into virtual audio device quirks/artifacts. CM is mostly useful for system-wide virtual surround and post-processing OpenAL effects)

> What prevents you from accepting calls from games coming to WASAPI and redirecting them from ASIO?
Just like Creative ALchemy relays DirectSound calls to OpenAL.

As far as I know, there's just no WASAPI to ASIO wrapper, like ALchemy/DSOAL can wrap DirectSound to OpenAL.
Something like https://github.com/kcat/openal-soft/issues/586 perhaps could work with an ASIO-capable OpenAL Soft build, but implementation documentation is mostly non-existent.
I think someone requested something like an WASAPI->OpenAL (Soft?) wrapper somewhere but I don't think it went anywhere and it's probably no easy task.
Which reminds me, some rare games may still use DirectSound (or offer it as an option) so you can also add WASAPI exclusive to those via DSOAL like I did here https://github.com/kcat/openal-soft/issues/682#issuecomment-2568690425

I'm not even a programmer like dechamps who's *literally* an ASIO developer, so he knows what he's talking about.
To put simply, there's just no way to do what you're asking that we know of. We suggested some alternatives with compromises because that's all we have at the moment. But anyway, I really dislike your arrogant attitude and also refuse to continue answering your barrage of baseless claims and [placebo expectations](https://github.com/user-attachments/assets/1fab6758-55df-4a46-9363-ad32cdd5fd70). Good luck✌

**Comment by MyGoodBear** [comment]
"I don't think that's what it really does"

I found out about this nonsense that I've never used and won't use 5 minutes ago.

I saw that it works with the OpenAL engine from Foobar and with ASIO - why it does this is unknown even to a spherical horse in a vacuum.

**Comment by MyGoodBear** [comment]
Here are the parameters of the game engine:

[WinDrv.WindowsClient]
AudioDeviceClass=ALAudio.ALAudioDevice

[ALAudio.ALAudioDevice]
UseEAX=True
Use3DSound=True
UseDefaultDriver=True
MaxEAXVersion=5
UsePrecache=True
MaxChannels=128
CompatibilityMode=False
UseFilter=True
UseReverb=True
TimeBetweenHWUpdates=15.000000
UseEffectsProcessing=True
LowQualitySound=False
DisablePitch=False
MinOggVorbisDurationGame=20
MinOggVorbisDurationEditor=4
MinOggVorbisDuration=10
DeviceName=SB X-Fi Audio [3FE0]
ReserveChannels=10
TotalAvailableChannels=128

From which it follows that the audio will be routed to the hardware device via OpenAL.

**Comment by MyGoodBear** [comment]
All modern sequencers are 64-bit - sound is loaded into them - it is output over ASIO, and in many channels.

Exactly the same thing should be done in all games, without exception.

**Comment by MyGoodBear** [comment]
<allowed value="software" />

**Comment by MyGoodBear** [comment]
"Which reminds me, some rare games may still use DirectSound (or offer it as an option) so you can also add WASAPI exclusive to those via DSOAL like I did here"

I don't need to do this with DirectSound games, as I have 3 Creative X-Fi hardware cards that can relay DirectSound to hardware OpenAL.

The problem is not with the old games, where everything is great, the problem is with all the new games that output sound to WASAPI, which destroys it.

**Comment by MyGoodBear** [comment]
"I'm not even a programmer like dechamps who's literally an ASIO developer, so he knows what he's talking about"

Developers and engineers are those who listen with devices through their eyes to the brain, and you need to listen with your ears to the brain.

Obviously.

**Comment by MyGoodBear** [comment]
Interview with Dave Ross Uma (EMU Company):

Maximum PC: But sound is a very subjective thing, most people won't be able to tell the difference.

Dave Rossum: Yes, the availability of education, or rather the lack of it, is a real problem.
The most important advice I can give is this: Listen to your equipment and follow your emotions. And whenever possible, standardize the tests you run.
Take a piece of music that you really like and listen to it from different places and with your eyes closed.
It will be great if you are satisfied with the sound of cheap equipment.
And if you like the sound of expensive equipment, and you can afford it, that's great too.

The only defining test is your own ears!
This is a completely subjective feeling.
Sound is an emotional part of a computer.
It can't be a reasonably explicable feeling.

Too often people try to use numerical measurements. I swear at my engineers every time they replace their ears with measuring devices. After all, as a result, they can create a circuit that sounds lousy, but shows good values on the devices.

![Image](https://github.com/user-attachments/assets/47e6c73d-0f09-420a-b718-48f44decc76e)

**Comment by MyGoodBear** [comment]
Needless to say, E-MU cards are some of the best-sounding cards?

![Image](https://github.com/user-attachments/assets/94b00c4d-e564-45ba-b3ef-850952d5346b)

![Image](https://github.com/user-attachments/assets/a24bb855-8483-4772-8ad8-222cf0f6cbc2)

**Comment by MyGoodBear** [comment]
Developers and engineers are those who must fulfill the understanding of those who hear, and should not dictate to those who hear what they hear or do not hear.
Everything is turned upside down.

I'm listening - You're writing code.

And I don't need to tell the person who writes the code what I hear and what I don't hear.

**Comment by MyGoodBear** [comment]
I don't go to the bakery and teach the baker how to bake rolls - he does it much better than me and understands much more about it.

**Comment by MyGoodBear** [comment]
There is a task:
Bitwise audio output from the game to the ASIO sound card machine.

As a programmer, I feel bad about doing this, and I won't teach him this, because he knows better how to do it.

I see the problem and set the task.

I can also check the quality of its execution by ear.

If it sounded, then the programmer is a great fellow, if not, there is no trial.

**Comment by MyGoodBear** [comment]
Where you see arrogance is setting a task in connection with an identified I/O problem that no one is able to solve.

Instead, you're all offering just fierce game that should never be used under any circumstances if you want to get high-quality Sound.

Don't suggest something that doesn't sound right, and understand the problem of audio input/output from modern games correctly, and no one will treat you "arrogantly."

**Comment by MyGoodBear** [comment]
No one is able to solve the problem, or even formulate it.

**Comment by LAGonauta** [comment]
Hi @MyGoodBear, I admire your passion on the subject but I don't think this is the proper platform for this discussion. Try the IRC channel or maybe even Discord if there is one.

**Comment by kcat** [comment]
> Developers and engineers are those who must fulfill the understanding of those who hear, and should not dictate to those who hear what they hear or do not hear.

When developers and engineers can verify the output is exactly the same as the input and show clear evidence that it is as they say, while "those who hear" are saying it's different and offer no actionable information, what can be done? That whole "Listen to your equipment and follow your emotions." is about being satisfied with what you have, don't go chasing ghosts thinking you need some prestigious high-end gear or some special setup as long as you're happy with what you got. It's not saying that if you feel it's not as good, then the system is objectively bad.

We know people have biased experiences, that two different people can be exposed to the same sound at the same time and come away with different feelings on how it sounded, and that people's hearing can be affected by a lot of external factors, such that the same person can perceive a difference with some sound despite it being exactly the same in all measurable aspects. This is why things like double-blind testing in controlled environments (to eliminate as many external factors as possible, including those that could affect us subconsciously), and in-depth measurements with meticulous records of the testing procedures, is important to know what's actually going on. We need more than "follow your emotions" to work with if we're to improve something.

In any case, as I mentioned in the other thread, OpenAL Soft has an option for WASAPI exclusive mode output. There is an in-progress-but-currently-unusable skeleton backend for ASIO output (which I may end up scrapping, since exclusive mode is supported and I hear ASIO can be annoyingly problematic due to device and driver quirks making things not function as they should). Other apps will depend on what options they have available and won't have a one-size-fits-all solution.

**Comment by MyGoodBear** [comment]
In general, to claim that a software mixer sounds the same as a hardware mixer when delivering data to it with bitwise precision is at least stupid and not sufficiently conscious.

Yes, WASAPI exclusive sounds better than the usual WASAPI due to the shorter signal transmission scheme, but it still doesn't sound like ASIO plays with bitwise precision, interacting with the card's device.

Especially when ASIO does not implement mixing from different programs.

If anything, there are two types of ASIO-enabled sound cards - these are ASIO-mixed cards and non-ASIO-mixed cards, and the latter cards sound better than ASIO-mixed cards.

That is, the nuances of sound recording and sound extraction are a wagon and a small cart, even when the numbers show you similar data when measuring.

Measure human emotions - I'll look at the result of these measurements and show you where you noticed them inaccurately.

**Comment by MyGoodBear** [comment]
"In any case, as I mentioned in the other thread, OpenAL Soft has an option for WASAPI exclusive mode output"


If you want a really good-sounding Soft, then you should output audio using ASIO or OpenAL hardware or BitStream.

I have never used and will not use the OpenAL software nickname, as I have heard how it sounds.
I have hardware OpenAL in three X-FI sound cards, and this sound is several orders of magnitude cooler than any soft mixing and signal transmission scheme.

Here on the channel, the highest quality audio available online on YouTube in 5.1 format.
https://youtu.be/7ds_Wl1rza0

You won't find a better sound on YouTube than this game in the world.

**Comment by MyGoodBear** [comment]
The key point is how the game transmits game audio data to the Windows mixer.

If this is something like a ready-made 7.1 circuit, then this data must be taken TO the Windows mixer and redirected channel-by-channel and bitwise to ASIO - this is the only way to get good Sound from modern games.

**Comment by MyGoodBear** [comment]
" I hear ASIO can be annoyingly problematic due to device and driver"

No. If you implement everything correctly and with straight hands (head), you can always achieve stable and good work.

To do this, there must be a parameter for changing the "Preload buffer", in addition to the ASIO buffer of the driver, and if the card is unable to support the reproducible frequency or bit rate, it is optional to offer a software frequency recalculation with a large red mark "Enabling these parameters spoils the Sound".

![Image](https://github.com/user-attachments/assets/376cc898-075b-4b8a-aceb-ebb1d8d7b6eb)

**Comment by MyGoodBear** [comment]
I literally just turned on the WASAPI exclusive I/O system many years later and compared it to ASIO, which I already know very well how it sounds.

https://youtu.be/CrTpF0LjoVE

1. They sound different.
I sang the song and monitored it with my own ears, with minimal delay.
Every experienced musician knows that in this mode, when you play music and monitor it, all the nuances and jambs of the monitoring system (input/output) are immediately audible.
2. So, firstly, WASAPI exclusive sounds more or less normal only at frequencies of 192 and 176.4 kHz - at the same time, the signal delay is minimal and generally acceptable for live performance.
At 96 and 88.2 kHz, the delay is already increasing, and with it the amount of "sand" in the sound, which I will talk about later.
At frequencies of 44.1 kHz and 48 kHz, it becomes impossible to sing due to the huge delay and the sound becomes even more "loose".
3. ASIO doesn't have all this at all, and it's not just about the delay.
What ASIO has at different frequencies during live performance is something that needs to be written about in a whole lecture, so I won't do it here and now.
4. And most importantly, What is different about the sound itself, if we ignore the difference in delays?
What is different is that there is some kind of "sand" in the sound of "WASAPI exclusive", which is clearly audible and expressed in the area of the Russian letters "H" and "S" - exactly that I I noticed it about 15 years ago, when I tested all this during stage performances at various events, as well as during the sound engineering of concerts by various bands.
There is some kind of "sand" in WASAPI exclusive - dirt that is impossible to get rid of.
Maybe it has something to do with something called "Dither":
https://youtu.be/zWpWIQw7HWU
With software noise generation, to save information when the bit rate changes.

I don't know what these distortions are related to, but they were in WASAPI exclusive 15 years ago and nothing has changed now - I still heard the same "loose" - "sandy" unpleasant sound.

ASIO doesn't have all this, it always sounds clean and clear - that's why sound engineers and musicians choose ASIO.

Which I always recommend to you.

**Comment by MyGoodBear** [comment]
Here is a demonstration of the obvious difference between the sound coming through the Windows mixer and bypassing it:
https://youtu.be/ROGBNc_rKhk

**Comment by MyGoodBear** [comment]
ASIO и OpenAL vs Windows WASAPI - for those who don't hear the obvious difference:
https://youtu.be/u3JaXiMsEK8

**Comment by MyGoodBear** [comment]
5.1 HardWare Mix VD HRTF - Binaural
https://youtu.be/X_5bpVj6mT0

**Comment by mirh** [comment]
For pete's sake, won't somebody clean this mess of a thread.

In other news @ThreeDeeJay can you double confirm your 3ms realtek results were done with their default drivers?
[People](https://forum.juce.com/t/juce-and-latency-on-windows-wasapi/35958/3) all over the place bitch they are lame.

**Comment by ThreeDeeJay** [comment]
@mirh Didn't you mean 10ms? https://github.com/kcat/openal-soft/issues/682#issuecomment-1144017398
I checked again and that's what I still get with my Realtek onboard audio, using the 2.82 driver from the official Realtek site. (2.83 is the latest but IIRC it broke 7.1 surround).
Or do you mean something else like WASAPI exclusive or ASIO? I forgot the details of this discussion lol
Either way, 2.82 is from like 2017 and I don't think I ever used anything else for long.

**Comment by mirh** [comment]
I meant https://github.com/kcat/openal-soft/issues/682#issuecomment-2568690425
And yes, with WASAPI exclusive, but still. Including the 160 samples thing that [somehow](https://forum.juce.com/t/enabling-windows-audios-src/41509/12) [nobody](https://forum.juce.com/t/juce-and-latency-on-windows-wasapi/35958/4) [could](https://forum.juce.com/t/windows-audio-low-latency-mode/42085/5) [get](https://stackoverflow.com/questions/37468283/how-to-get-below-10ms-latency-using-wasapi-shared-mode#comment65476174_37468283) [working](https://eab.abime.net/showthread.php?t=83945#post1115506).

**Comment by ThreeDeeJay** [comment]
Oh I forgot about that and couldn't find it because for some reason my benchmark messages were hidden. 🤔 

Anyhow, here's the driver I'm using:
<img width="364" height="210" alt="Image" src="https://github.com/user-attachments/assets/0de53df1-6372-4bf5-a911-e0afce60df39" />

And I just tested FreePiano again with DSOAL r689 + OpenAL Soft r10273
<img width="333" height="453" alt="Image" src="https://github.com/user-attachments/assets/20b091cc-791c-4ab1-8f93-e2d90e189c8a" /> <img width="333" height="456" alt="Image" src="https://github.com/user-attachments/assets/4244bf00-a96e-41b7-97af-94194d200dc1" />

So I still got 3.6ms with `exclusive-mode=true` when `period_size` is set to anywhere from `1` to `174`  even though `alsoft_error.txt` reports `64 / 192 buffer` and `174 / 522 buffer`, respectively. https://www.diffchecker.com/uZ37hC4A/
and as soon as I set it to `175`, it goes up to 4.4ms. And oddly, different `periods` values don't make FreePiano report different latency, so again, I'm still not quite sure how it measures latency or how accurate it is.
